#33903: Flexible KernelDescriptor runtime args

ssundaramTT opened the pull request:
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/33903)

### Problem description
Currently, runtime args is a triple-nested vector where `rtargs[i][j]` is a vector of runtime args for core (i, j). If a kernel does not require arguments for all cores starting from (0, 0), it is awkward to populate with a bunch of blank vectors.

### What's changed

- Changed runtime args to be `std::vector<std::pair<CoreCoord, std::vector<uint32_t>>>`
- Updated generic op program factory and test file accordingly
- [Gist for how to use new API](https://gist.github.com/ssundaramTT/e85e33f66f1e83d8fde6f4f275bfa238)

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes ([link to run](https://github.com/tenstorrent/tt-metal/actions/runs/20446009771))
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

ssundaramTT committed:
changed rtargs from vector of vectors to vector of pairs

ssundaramTT committed:
applied changes to generic op test

ssundaramTT committed:
Merge branch 'main' into ssundaram/kernel_rtargs

ssundaramTT committed:
precommit hooks

akerteszTT commented on tt_metal/api/tt-metalium/program_descriptors.hpp:
```diff
@@ -102,7 +102,7 @@ struct KernelDescriptor {
     using Defines = std::vector<std::pair<std::string, std::string>>;
     using CoreRuntimeArgs = std::vector<uint32_t>;
-    using RuntimeArgs = std::vector<std::vector<CoreRuntimeArgs>>;
+    using RuntimeArgs = std::vector<std::pair<CoreCoord, CoreRuntimeArgs>>;
```
This is an API-breaking change.

However, since you're modifying a public struct, it is _very_ difficult to apply the 2-step deprecation of our API stability policy. You'd have to make a "V2" of the struct, and temporarily duplicate APIs that accept it...

akerteszTT reviewed (approved):
This PR constitutes an API-breaking change.

Given the extremely narrow usage, and the difficulty of applying deprecation for this type of change, this can be a stability exception.

ssundaramTT committed:
maintain python indexing

ssundaramTT committed:
Merge branch 'main' into ssundaram/kernel_rtargs

ssundaramTT committed:
Add RuntimeArgsWrapper to nanobind

Add RuntimeArgsWrapper and RuntimeArgsRowProxy helper classes to the
nanobind implementation. This enables Python 2D indexing syntax
(rtargs[i][j] = [...]) for kernel runtime arguments.

Note: The pybind11 version is frozen and unchanged per project policy.

ssundaramTT committed:
Revert pybind changes - pybind11 bindings are frozen

Remove RuntimeArgsWrapper from pybind file as pybind11 bindings are
deprecated and frozen per project policy. The RuntimeArgsWrapper
functionality is available in the nanobind implementation only.

ssundaramTT committed:
Fix RuntimeArgs usage in test_generic_op

Use ttnn.RuntimeArgs() wrapper class instead of nested Python lists.
The wrapper enables 2D indexing syntax: rtargs[y][x] = args creates
CoreCoord(x, y) with the specified runtime arguments.

ssundaramTT committed:
wrapper class name change

ssundaramTT committed:
clang tidy
